src/OpenTelemetry.Instrumentation.StackExchangeRedis/StackExchangeRedisInstrumentation.cs

```diff
@@ -0,0 +1,107 @@
+// <copyright file="StackExchangeRedisInstrumentation.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using Microsoft.Extensions.Options;
+using OpenTelemetry.Internal;
+using StackExchange.Redis;
+
+namespace OpenTelemetry.Instrumentation.StackExchangeRedis;
+
+/// <summary>
+/// StackExchange.Redis instrumentation.
+/// </summary>
+public sealed class StackExchangeRedisInstrumentation : IDisposable
+{
+    private readonly IOptionsMonitor<StackExchangeRedisInstrumentationOptions> options;
+
+    internal StackExchangeRedisInstrumentation(
+        IOptionsMonitor<StackExchangeRedisInstrumentationOptions> options)
+    {
+        this.options = options;
+    }
+
+    internal List<StackExchangeRedisConnectionInstrumentation> InstrumentedConnections { get; } = new();
+
+    /// <summary>
+    /// Adds an <see cref="IConnectionMultiplexer"/> to the instrumentation.
+    /// </summary>
+    /// <param name="connection"><see cref="IConnectionMultiplexer"/>.</param>
+    /// <returns><see cref="IDisposable"/> to cancel the registration.</returns>
+    public IDisposable AddConnection(IConnectionMultiplexer connection)
+        => this.AddConnection(Options.DefaultName, connection);
+
+    /// <summary>
+    /// Adds an <see cref="IConnectionMultiplexer"/> to the instrumentation.
+    /// </summary>
+    /// <param name="name">Name to use when retrieving options.</param>
+    /// <param name="connection"><see cref="IConnectionMultiplexer"/>.</param>
+    /// <returns><see cref="IDisposable"/> to cancel the registration.</returns>
+    public IDisposable AddConnection(string name, IConnectionMultiplexer connection)
+    {
+        Guard.ThrowIfNull(name);
+        Guard.ThrowIfNull(connection);
+
+        var options = this.options.Get(name);
+
+        lock (this.InstrumentedConnections)
+        {
+            var instrumentation = new StackExchangeRedisConnectionInstrumentation(connection, options);
```

Comment:
Do we want to allow multiple connections?  If yes, we should probably update the thread name to be distinct when instantiating multiple instances of `StackExchangeRedisConnectionInstrumentation`.

Reply:
When our application calls `AddRedisInstrumentation` multiple times (to instrument different connections), we get duplicate spans in our traces (one for each instrumented connection even though the Redis call is only being directed to one of those connections). Is there a way to avoid the duplicate spans?

Seeing this duplicate span behavior so far in the following versions:
- `OpenTelemetry.Instrumentation.StackExchangeRedis v1.0.0-rc9.9`
- `OpenTelemetry.Instrumentation.StackExchangeRedis v1.0.0-rc9.8`
- `OpenTelemetry.Instrumentation.StackExchangeRedis v1.0.0-rc9.7`

Can continue testing up to the current version if that helps.

I submitted this PR (https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/2001) in an attempt to address what we are seeing in our application, but the changes haven't been well received, so they may not be merged.

Open to other solutions, but so far coming up empty.